tests/rados/stretch_cluster.py

```diff
@@ -0,0 +1,472 @@
+import datetime
+import json
+import logging
+import re
+import time
+
+from ceph.ceph_admin import CephAdmin
+from ceph.parallel import parallel
+from tests.rados.mute_alerts import get_alerts
+from tests.rados.rados_prep import create_pool
+from tests.rados.test_9281 import do_rados_get, do_rados_put
+
+log = logging.getLogger(__name__)
+
+
+def run(ceph_cluster, **kw):
+    """
+    enables connectivity mode and deploys stretch cluster with arbiter mon node
+    Actions Performed:
+    1. Disables the automatic crush map update
+    2. Collects the OSD daemons in the cluster and split them into 2 sites.
+    3. If add capacity is selected, only half of the OSD's will be added to various sites initially.
+    4. Adds the stretch rule into crush map.
+    5. Adding monitors into the 2 sites.
+    6. Create a replicated pool and deploy stretch mode.
+    7. Create a test pool, write some data and perform add capacity. ( add osd nodes into two sites )
+    8. Check for the bump in election epochs throughout.
+    9. Check the acting set in PG for 4 OSD's. 2 from each site.
+    Args:
+        ceph_cluster (ceph.ceph.Ceph): ceph cluster
+    """
+    log.info("Deploying stretch cluster with arbiter mon node")
+    log.info(run.__doc__)
+    config = kw.get("config")
+    cephadm = CephAdmin(cluster=ceph_cluster, **config)
+    client_node = None
+    tiebreaker_node = None
+    ceph_nodes = kw.get("ceph_nodes")
+    for node in ceph_nodes:
+        if node.role == "installer":
+            tiebreaker_node = node
+        if node.role == "client":
+            client_node = node
+
+    if not client_node and not tiebreaker_node:
+        log.error(
+            "Admin client and tie breaker node not configured, Cannot modify crush rules for stretch cluster"
+        )
+        return 1
+
+    mon_state = get_mon_details(node=cephadm)
+    if len(list(mon_state["monitors"])) < 5:
+        log.error(
+            f"Minimum of 5 Mon daemons needed to deploy a stretch cluster, found : {len(mon_state['monitors'])}"
+        )
+        return 1
+    osd_details = get_osd_details(node=cephadm)
+    if len(osd_details.keys()) < 4:
+        log.error(
+            f"Minimum of 4 osd daemons needed to deploy a stretch cluster, found : {len(osd_details.keys())}"
+        )
+        return 1
+
+    # disabling automatic crush update
+    cmd = "ceph config set osd osd_crush_update_on_start false"
+    cephadm.shell([cmd])
```

Comment:
Would it make sense to use `exec_command` instead of `shell`? As the client is setup with the `admin.keyring`, I believe it is possible to use `client.exec_command(cmd=cmd)`

Reply:
TO keep the function call uniform, I used `shell`. and used `exec_command` only in method to add the crush rules. Yea, it is possible to perform those operations from client.